{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to the SQL database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'duckdb'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mduckdb\u001b[39;00m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mpandas\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mpd\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mipywidgets\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'duckdb'"
     ]
    }
   ],
   "source": [
    "import duckdb\n",
    "import pandas as pd\n",
    "import ipywidgets\n",
    "\n",
    "%reload_ext sql\n",
    "\n",
    "%config SqlMagic.autopandas = True\n",
    "%config SqlMagic.feedback = False\n",
    "%config SqlMagic.displaycon = False\n",
    "\n",
    "%sql duckdb:///vcf_dfs.db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "%%sql\n",
    "SET memory_limit = '10GB';\n",
    "SET enable_progress_bar = true;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the locations of the references and samples, and quality/depth cutoffs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_parent_file = \"../../data/parquets/w1118.parquet\"\n",
    "alt_parent_file = \"../../data/parquets/oregonr.parquet\"\n",
    "sample_file_glob = \"../../data/parquets/WT-G0-*.parquet\"\n",
    "ref_parent_name = \"w1118\"\n",
    "alt_parent_name = \"oregonr\"\n",
    "\n",
    "ref_qual_cutoff = 200\n",
    "ref_depth_cutoff = 30\n",
    "\n",
    "ref_out_file = \"../../data/parquets/reference.parquet\"\n",
    "sample_out_file = \"../../data/parquets/progeny.parquet\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import the reference files as a table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "RuntimeError: (duckdb.duckdb.BinderException) Binder Error: GROUP BY clause cannot contain aggregates!\n",
      "LINE 2: GROUP BY ANY_VALUE(sample), chromosome, int_pos;\n",
      "                 ^\n",
      "[SQL: SELECT * FROM parents\n",
      "GROUP BY ANY_VALUE(sample), chromosome, int_pos;]\n",
      "(Background on this error at: https://sqlalche.me/e/20/f405)\n",
      "If you need help solving this issue, send us a message: https://ploomber.io/community\n"
     ]
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE OR REPLACE VIEW parents AS\n",
    "SELECT sample, chromosome, CAST(position AS INTEGER) AS int_pos, quality, genotype, depth, allele_depth, reference, variant, (CASE WHEN variant='.' THEN reference ELSE variant END) AS mod_variant FROM read_parquet(['{{ref_parent_file}}', '{{alt_parent_file}}']);\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Isolate sites where there are 2 variants (meaning one parent is different from the other)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE OR REPLACE VIEW check_unique_variants AS\n",
    "SELECT chromosome, int_pos, COUNT(variant) AS n_variants FROM parents\n",
    "    GROUP BY chromosome, int_pos\n",
    "    HAVING n_variants >= 2;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create new reference from the reference parent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "31d138ee147d4a4b956c4ab811ed427f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>int_pos</th>\n",
       "      <th>reference</th>\n",
       "      <th>mod_variant</th>\n",
       "      <th>quality</th>\n",
       "      <th>genotype</th>\n",
       "      <th>depth</th>\n",
       "      <th>allele_depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>13247</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>16</td>\n",
       "      <td>0,16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>13638</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>222.000000</td>\n",
       "      <td>1/1</td>\n",
       "      <td>13</td>\n",
       "      <td>0,13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>13638</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>31</td>\n",
       "      <td>0,31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>20008</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>20</td>\n",
       "      <td>0,20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>28813</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>17</td>\n",
       "      <td>0,17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276434</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chrY</td>\n",
       "      <td>975741</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>175</td>\n",
       "      <td>7,168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276435</th>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chrY</td>\n",
       "      <td>975741</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>275</td>\n",
       "      <td>26,249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276436</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chrY</td>\n",
       "      <td>975783</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>171</td>\n",
       "      <td>3,168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276437</th>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chrY</td>\n",
       "      <td>975783</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>263</td>\n",
       "      <td>8,255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>276438</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chrY</td>\n",
       "      <td>2777139</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>209.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>28</td>\n",
       "      <td>2,26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>276439 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  sample chromosome  int_pos reference mod_variant  \\\n",
       "0         w1118_combined      chr2L    13247         C           T   \n",
       "1         w1118_combined      chr2L    13638         G           A   \n",
       "2       oregonr_combined      chr2L    13638         G           A   \n",
       "3         w1118_combined      chr2L    20008         A           G   \n",
       "4         w1118_combined      chr2L    28813         A           G   \n",
       "...                  ...        ...      ...       ...         ...   \n",
       "276434    w1118_combined       chrY   975741         A           T   \n",
       "276435  oregonr_combined       chrY   975741         A           T   \n",
       "276436    w1118_combined       chrY   975783         T           A   \n",
       "276437  oregonr_combined       chrY   975783         T           A   \n",
       "276438    w1118_combined       chrY  2777139         T           G   \n",
       "\n",
       "           quality genotype  depth allele_depth  \n",
       "0       221.998993      1/1     16         0,16  \n",
       "1       222.000000      1/1     13         0,13  \n",
       "2       221.998993      1/1     31         0,31  \n",
       "3       221.998993      1/1     20         0,20  \n",
       "4       221.998993      1/1     17         0,17  \n",
       "...            ...      ...    ...          ...  \n",
       "276434  221.998993      1/1    175        7,168  \n",
       "276435  221.998993      1/1    275       26,249  \n",
       "276436  221.998993      1/1    171        3,168  \n",
       "276437  221.998993      1/1    263        8,255  \n",
       "276438  209.998993      1/1     28         2,26  \n",
       "\n",
       "[276439 rows x 9 columns]"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE OR REPLACE VIEW temp_ref AS\n",
    "SELECT\n",
    "    sample,\n",
    "    parents.chromosome,\n",
    "    parents.int_pos,\n",
    "    reference,\n",
    "    mod_variant,\n",
    "    quality,\n",
    "    genotype,\n",
    "    depth,\n",
    "    allele_depth\n",
    "    FROM parents\n",
    "    INNER JOIN check_unique_variants ON parents.chromosome = check_unique_variants.chromosome AND parents.int_pos = check_unique_variants.int_pos\n",
    "    WHERE genotype!='0/1'\n",
    "        AND quality > '{{ref_qual_cutoff}}'\n",
    "        AND LENGTH(mod_variant) <= 1\n",
    "        AND LENGTH(reference) <= 1\n",
    "        AND depth > '{{ref_depth_cutoff}}';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cf6f3cf6877f4267a58370002039cda2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>int_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>chromosome_1</th>\n",
       "      <th>int_pos_1</th>\n",
       "      <th>alt_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>13247.0</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>13638.0</td>\n",
       "      <td>A</td>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>13638.0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>20008.0</td>\n",
       "      <td>G</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>28813.0</td>\n",
       "      <td>G</td>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>28813.0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>w1118_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>36330.0</td>\n",
       "      <td>T</td>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>36330.0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175393</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>286991.0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175394</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>65405.0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175395</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>16119415.0</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175396</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>16126653.0</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>175397</th>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>oregonr_combined</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>22791947.0</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>175398 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                sample chromosome  int_pos ref_allele          sample_1  \\\n",
       "0       w1118_combined      chr2L  13247.0          T              None   \n",
       "1       w1118_combined      chr2L  13638.0          A  oregonr_combined   \n",
       "2       w1118_combined      chr2L  20008.0          G              None   \n",
       "3       w1118_combined      chr2L  28813.0          G  oregonr_combined   \n",
       "4       w1118_combined      chr2L  36330.0          T  oregonr_combined   \n",
       "...                ...        ...      ...        ...               ...   \n",
       "175393            None       None      NaN       None  oregonr_combined   \n",
       "175394            None       None      NaN       None  oregonr_combined   \n",
       "175395            None       None      NaN       None  oregonr_combined   \n",
       "175396            None       None      NaN       None  oregonr_combined   \n",
       "175397            None       None      NaN       None  oregonr_combined   \n",
       "\n",
       "       chromosome_1   int_pos_1 alt_allele  \n",
       "0              None         NaN       None  \n",
       "1             chr2L     13638.0          A  \n",
       "2              None         NaN       None  \n",
       "3             chr2L     28813.0          G  \n",
       "4             chr2L     36330.0          T  \n",
       "...             ...         ...        ...  \n",
       "175393        chr2L    286991.0          A  \n",
       "175394        chr2L     65405.0          T  \n",
       "175395        chr2R  16119415.0          G  \n",
       "175396        chr2R  16126653.0          A  \n",
       "175397        chr2R  22791947.0          T  \n",
       "\n",
       "[175398 rows x 8 columns]"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "CREATE OR REPLACE VIEW ref_parent AS\n",
    "SELECT sample, chromosome, int_pos, mod_variant AS ref_allele FROM temp_ref\n",
    "WHERE sample='{{ref_parent_name}}';\n",
    "\n",
    "CREATE OR REPLACE VIEW alt_parent AS\n",
    "SELECT sample, chromosome, int_pos, mod_variant AS alt_allele FROM temp_ref\n",
    "WHERE sample='{{alt_parent_name}}';\n",
    "\n",
    "CREATE OR REPLACE TABLE ref AS\n",
    "SELECT * FROM ref_parent\n",
    "FULL JOIN alt_parent ON (ref_parent.chromosome = alt_parent.chromosome AND ref_parent.int_pos = alt_parent.int_pos)\n",
    "ORDER BY ref_parent.chromosome, ref_parent.int_pos;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "707d6f6848d54e40be155166bda37d2c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "%%sql\n",
    "\n",
    "CREATE OR REPLACE VIEW temp_vcfs AS\n",
    "SELECT * FROM read_parquet('{{sample_file_glob}}');\n",
    "\n",
    "CREATE OR REPLACE VIEW vcfs AS\n",
    "SELECT\n",
    "    sample,\n",
    "    chromosome,\n",
    "    CAST(position AS INTEGER) AS int_pos,\n",
    "    reference,\n",
    "    variant,\n",
    "    quality,\n",
    "    genotype,\n",
    "    depth,\n",
    "    allele_depth,\n",
    "    (CASE WHEN temp_vcfs.variant='.' THEN temp_vcfs.reference ELSE temp_vcfs.variant END) AS new_variant\n",
    "    FROM temp_vcfs;\n",
    "\n",
    "CREATE OR REPLACE VIEW samples_rearranged AS\n",
    "SELECT *, (CASE WHEN ref_allele=new_variant THEN '.' ELSE new_variant END) AS var_adjusted FROM vcfs\n",
    "    INNER JOIN ref ON vcfs.chromosome = ref.chromosome AND vcfs.int_pos = ref.int_pos\n",
    "    WHERE new_variant=ref_allele OR new_variant=alt_allele;\n",
    "\n",
    "CREATE OR REPLACE TABLE final_samples AS\n",
    "SELECT \n",
    "    string_split(sample, '-')[1] AS condition,\n",
    "    string_split(sample, '-')[2] AS sample_type,\n",
    "    string_split(sample, '-')[3] AS sample_num,\n",
    "    reference,\n",
    "    variant,\n",
    "    chromosome,\n",
    "    int_pos AS position,\n",
    "    string_split(allele_depth, ',')[1] AS ref_reads,\n",
    "    string_split(allele_depth, ',')[2] AS variant_reads,\n",
    "    quality AS QUAL,\n",
    "    genotype AS GT,\n",
    "    depth AS DP\n",
    "    FROM samples_rearranged\n",
    "ORDER BY sample_num, chromosome, position;\n",
    "\n",
    "CHECKPOINT;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sample_num</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_reads</th>\n",
       "      <th>variant_reads</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>GT</th>\n",
       "      <th>DP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>001</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>1023412</td>\n",
       "      <td>0</td>\n",
       "      <td>87</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>001</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>1150444</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>001</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>1617652</td>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>001</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>1712703</td>\n",
       "      <td>0</td>\n",
       "      <td>70</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>001</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>1754752</td>\n",
       "      <td>0</td>\n",
       "      <td>75</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1448</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>016</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr3R</td>\n",
       "      <td>30001926</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1449</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>016</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr3R</td>\n",
       "      <td>31655172</td>\n",
       "      <td>26</td>\n",
       "      <td>20</td>\n",
       "      <td>192.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1450</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>016</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chrX</td>\n",
       "      <td>10005635</td>\n",
       "      <td>46</td>\n",
       "      <td>59</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1451</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>016</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chrX</td>\n",
       "      <td>15677336</td>\n",
       "      <td>0</td>\n",
       "      <td>152</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1452</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>016</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chrX</td>\n",
       "      <td>19521073</td>\n",
       "      <td>0</td>\n",
       "      <td>99</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1453 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     condition sample_type sample_num reference variant chromosome  position  \\\n",
       "0           WT          G0        001         G       A      chr2L   1023412   \n",
       "1           WT          G0        001         C       T      chr2L   1150444   \n",
       "2           WT          G0        001         A       C      chr2L   1617652   \n",
       "3           WT          G0        001         G       A      chr2L   1712703   \n",
       "4           WT          G0        001         G       C      chr2L   1754752   \n",
       "...        ...         ...        ...       ...     ...        ...       ...   \n",
       "1448        WT          G0        016         C       A      chr3R  30001926   \n",
       "1449        WT          G0        016         T       G      chr3R  31655172   \n",
       "1450        WT          G0        016         T       G       chrX  10005635   \n",
       "1451        WT          G0        016         A       C       chrX  15677336   \n",
       "1452        WT          G0        016         A       T       chrX  19521073   \n",
       "\n",
       "     ref_reads variant_reads        QUAL   GT   DP  \n",
       "0            0            87  221.998993  1/1   87  \n",
       "1            0            88  221.998993  1/1   88  \n",
       "2            0            28  221.998993  1/1   28  \n",
       "3            0            70  221.998993  1/1   70  \n",
       "4            0            75  221.998993  1/1   75  \n",
       "...        ...           ...         ...  ...  ...  \n",
       "1448         0            97  221.998993  1/1   97  \n",
       "1449        26            20  192.009003  0/1   46  \n",
       "1450        46            59  225.009003  0/1  105  \n",
       "1451         0           152  221.998993  1/1  152  \n",
       "1452         0            99  221.998993  1/1   99  \n",
       "\n",
       "[1453 rows x 12 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "COPY (SELECT * FROM final_samples)\n",
    "TO '{{ref_out_file}}'\n",
    "(FORMAT 'parquet');\n",
    "\n",
    "COPY (SELECT chromosome, int_pos AS position, ref_allele AS reference, alt_allele AS variant FROM ref)\n",
    "TO '{{sample_out_file}}'\n",
    "(FORMAT 'parquet');\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
