{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Connect to the SQL database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import duckdb\n",
    "import pandas as pd\n",
    "import ipywidgets\n",
    "\n",
    "%reload_ext sql\n",
    "\n",
    "%config SqlMagic.autopandas = True\n",
    "%config SqlMagic.feedback = False\n",
    "%config SqlMagic.displaycon = False\n",
    "\n",
    "%sql duckdb:///vcf_dfs2.db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SET enable_progress_bar = true;\n",
    "SET preserve_insertion_order = false;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Define the locations of the references and samples, and quality/depth cutoffs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_parent_file = \"../../data/parquets/w1118.parquet\"\n",
    "alt_parent_file = \"../../data/parquets/oregonr.parquet\"\n",
    "sample_file_glob = \"../../data/parquets/WT-G0-0*.parquet\"\n",
    "ref_parent_name = \"w1118\"\n",
    "alt_parent_name = \"oregonr\"\n",
    "\n",
    "ref_qual_cutoff = 200\n",
    "ref_depth_cutoff = 15\n",
    "\n",
    "ref_out_file = \"../../data/parquets/reference.parquet2\"\n",
    "sample_out_file = \"../../data/parquets/progeny.parquet2\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Import the reference files as a table. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE OR REPLACE VIEW parents AS\n",
    "SELECT sample, chromosome, CAST(position AS INTEGER) AS int_pos, quality, genotype, depth, allele_depth, UPPER(reference) AS reference, UPPER(variant) AS variant, (CASE WHEN variant='.' THEN reference ELSE UPPER(variant) END) AS mod_variant FROM read_parquet(['{{ref_parent_file}}', '{{alt_parent_file}}']);\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Isolate sites where there are 2 variants (meaning one parent is different from the other)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE OR REPLACE VIEW check_unique_variants AS\n",
    "SELECT chromosome, int_pos, COUNT(variant) AS n_variants FROM parents\n",
    "    GROUP BY chromosome, int_pos\n",
    "    HAVING n_variants >= 2;"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create new reference from the reference parent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE OR REPLACE VIEW temp_ref AS\n",
    "SELECT\n",
    "    sample,\n",
    "    parents.chromosome,\n",
    "    parents.int_pos,\n",
    "    reference,\n",
    "    mod_variant,\n",
    "    quality,\n",
    "    genotype,\n",
    "    depth,\n",
    "    allele_depth\n",
    "    FROM parents\n",
    "    INNER JOIN check_unique_variants ON parents.chromosome = check_unique_variants.chromosome AND parents.int_pos = check_unique_variants.int_pos\n",
    "    WHERE genotype!='0/1'\n",
    "        AND quality > '{{ref_qual_cutoff}}'\n",
    "        AND LENGTH(mod_variant) <= 1\n",
    "        AND LENGTH(reference) <= 1\n",
    "        AND depth > '{{ref_depth_cutoff}}';"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3596aa90d9df4b68b3ad91dd5415ee60",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "CREATE OR REPLACE VIEW ref_parent AS\n",
    "SELECT sample, chromosome, int_pos, mod_variant AS ref_allele FROM temp_ref\n",
    "WHERE sample='{{ref_parent_name}}';\n",
    "\n",
    "CREATE OR REPLACE VIEW alt_parent AS\n",
    "SELECT sample, chromosome, int_pos, mod_variant AS alt_allele FROM temp_ref\n",
    "WHERE sample='{{alt_parent_name}}';\n",
    "\n",
    "CREATE OR REPLACE TABLE ref AS\n",
    "SELECT * FROM ref_parent\n",
    "INNER JOIN alt_parent ON (ref_parent.chromosome = alt_parent.chromosome AND ref_parent.int_pos = alt_parent.int_pos)\n",
    "ORDER BY ref_parent.chromosome, ref_parent.int_pos;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "CREATE OR REPLACE TABLE refs_unique AS\n",
    "SELECT * FROM ref\n",
    "WHERE ref_allele != alt_allele;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>int_pos</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>chromosome_1</th>\n",
       "      <th>int_pos_1</th>\n",
       "      <th>alt_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13843951</td>\n",
       "      <td>G</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13843951</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13843981</td>\n",
       "      <td>A</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13843981</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13845603</td>\n",
       "      <td>A</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13845603</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13845663</td>\n",
       "      <td>C</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13845663</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13845674</td>\n",
       "      <td>A</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>13845674</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83367</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10002638</td>\n",
       "      <td>G</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10002638</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83368</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10002726</td>\n",
       "      <td>G</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10002726</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83369</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10002727</td>\n",
       "      <td>C</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10002727</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83370</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10003359</td>\n",
       "      <td>A</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10003359</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83371</th>\n",
       "      <td>w1118</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10003510</td>\n",
       "      <td>G</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr2R</td>\n",
       "      <td>10003510</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>83372 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample chromosome   int_pos ref_allele sample_1 chromosome_1  int_pos_1  \\\n",
       "0      w1118      chr2R  13843951          G  oregonr        chr2R   13843951   \n",
       "1      w1118      chr2R  13843981          A  oregonr        chr2R   13843981   \n",
       "2      w1118      chr2R  13845603          A  oregonr        chr2R   13845603   \n",
       "3      w1118      chr2R  13845663          C  oregonr        chr2R   13845663   \n",
       "4      w1118      chr2R  13845674          A  oregonr        chr2R   13845674   \n",
       "...      ...        ...       ...        ...      ...          ...        ...   \n",
       "83367  w1118      chr2R  10002638          G  oregonr        chr2R   10002638   \n",
       "83368  w1118      chr2R  10002726          G  oregonr        chr2R   10002726   \n",
       "83369  w1118      chr2R  10002727          C  oregonr        chr2R   10002727   \n",
       "83370  w1118      chr2R  10003359          A  oregonr        chr2R   10003359   \n",
       "83371  w1118      chr2R  10003510          G  oregonr        chr2R   10003510   \n",
       "\n",
       "      alt_allele  \n",
       "0              A  \n",
       "1              T  \n",
       "2              T  \n",
       "3              T  \n",
       "4              G  \n",
       "...          ...  \n",
       "83367          C  \n",
       "83368          C  \n",
       "83369          T  \n",
       "83370          C  \n",
       "83371          A  \n",
       "\n",
       "[83372 rows x 8 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT * FROM refs_unique;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "CREATE OR REPLACE VIEW temp_vcfs AS\n",
    "SELECT * FROM read_parquet('{{sample_file_glob}}');"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "CREATE OR REPLACE VIEW vcfs AS\n",
    "SELECT\n",
    "    sample,\n",
    "    chromosome,\n",
    "    CAST(position AS INTEGER) AS int_pos,\n",
    "    reference,\n",
    "    variant,\n",
    "    quality,\n",
    "    genotype,\n",
    "    depth,\n",
    "    allele_depth,\n",
    "    (CASE WHEN temp_vcfs.variant='.' THEN temp_vcfs.reference ELSE temp_vcfs.variant END) AS new_variant\n",
    "    FROM temp_vcfs;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "CREATE OR REPLACE VIEW samples_rearranged AS\n",
    "SELECT *, (CASE WHEN ref_allele=new_variant THEN '.' ELSE new_variant END) AS var_adjusted FROM vcfs\n",
    "    INNER JOIN refs_unique ON vcfs.chromosome = refs_unique.chromosome AND vcfs.int_pos = refs_unique.int_pos\n",
    "    WHERE new_variant=ref_allele OR new_variant=alt_allele;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6bd076ad053e4e70ac5d0db53b5bcb05",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dc63abd7375343a798399ae56db7b662",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>int_pos</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>quality</th>\n",
       "      <th>genotype</th>\n",
       "      <th>depth</th>\n",
       "      <th>allele_depth</th>\n",
       "      <th>new_variant</th>\n",
       "      <th>sample_1</th>\n",
       "      <th>chromosome_1</th>\n",
       "      <th>int_pos_1</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>sample_1_1</th>\n",
       "      <th>chromosome_1_1</th>\n",
       "      <th>int_pos_1_1</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>var_adjusted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>655648</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>86</td>\n",
       "      <td>38,48</td>\n",
       "      <td>T</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>655648</td>\n",
       "      <td>G</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>655648</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>661618</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>240.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>70</td>\n",
       "      <td>70</td>\n",
       "      <td>T</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>661618</td>\n",
       "      <td>t</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>661618</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>660851</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>69</td>\n",
       "      <td>34,35</td>\n",
       "      <td>A</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>660851</td>\n",
       "      <td>G</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>660851</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>668738</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>93.007698</td>\n",
       "      <td>0/1</td>\n",
       "      <td>51</td>\n",
       "      <td>41,10</td>\n",
       "      <td>A</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>668738</td>\n",
       "      <td>C</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>668738</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>694885</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>73</td>\n",
       "      <td>41,32</td>\n",
       "      <td>G</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>694885</td>\n",
       "      <td>A</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>694885</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>699318</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>84</td>\n",
       "      <td>45,39</td>\n",
       "      <td>C</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>699318</td>\n",
       "      <td>T</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>699318</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>714112</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>82</td>\n",
       "      <td>36,46</td>\n",
       "      <td>A</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>714112</td>\n",
       "      <td>G</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>714112</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>714319</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>72</td>\n",
       "      <td>27,45</td>\n",
       "      <td>G</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>714319</td>\n",
       "      <td>T</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>714319</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>749771</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>76</td>\n",
       "      <td>34,42</td>\n",
       "      <td>T</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>749771</td>\n",
       "      <td>C</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>749771</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>749936</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>225.009003</td>\n",
       "      <td>0/1</td>\n",
       "      <td>62</td>\n",
       "      <td>33,29</td>\n",
       "      <td>T</td>\n",
       "      <td>w1118</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>749936</td>\n",
       "      <td>A</td>\n",
       "      <td>oregonr</td>\n",
       "      <td>chr3L</td>\n",
       "      <td>749936</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample chromosome  int_pos reference variant     quality genotype  \\\n",
       "0  WT-G0-001      chr3L   655648         G       T  225.009003      0/1   \n",
       "1  WT-G0-001      chr3L   661618         T       .  240.996002      0/0   \n",
       "2  WT-G0-001      chr3L   660851         G       A  225.009003      0/1   \n",
       "3  WT-G0-001      chr3L   668738         C       A   93.007698      0/1   \n",
       "4  WT-G0-001      chr3L   694885         A       G  225.009003      0/1   \n",
       "5  WT-G0-001      chr3L   699318         T       C  225.009003      0/1   \n",
       "6  WT-G0-001      chr3L   714112         G       A  225.009003      0/1   \n",
       "7  WT-G0-001      chr3L   714319         T       G  225.009003      0/1   \n",
       "8  WT-G0-001      chr3L   749771         C       T  225.009003      0/1   \n",
       "9  WT-G0-001      chr3L   749936         A       T  225.009003      0/1   \n",
       "\n",
       "   depth allele_depth new_variant sample_1 chromosome_1  int_pos_1 ref_allele  \\\n",
       "0     86        38,48           T    w1118        chr3L     655648          G   \n",
       "1     70           70           T    w1118        chr3L     661618          t   \n",
       "2     69        34,35           A    w1118        chr3L     660851          G   \n",
       "3     51        41,10           A    w1118        chr3L     668738          C   \n",
       "4     73        41,32           G    w1118        chr3L     694885          A   \n",
       "5     84        45,39           C    w1118        chr3L     699318          T   \n",
       "6     82        36,46           A    w1118        chr3L     714112          G   \n",
       "7     72        27,45           G    w1118        chr3L     714319          T   \n",
       "8     76        34,42           T    w1118        chr3L     749771          C   \n",
       "9     62        33,29           T    w1118        chr3L     749936          A   \n",
       "\n",
       "  sample_1_1 chromosome_1_1  int_pos_1_1 alt_allele var_adjusted  \n",
       "0    oregonr          chr3L       655648          T            T  \n",
       "1    oregonr          chr3L       661618          T            T  \n",
       "2    oregonr          chr3L       660851          A            A  \n",
       "3    oregonr          chr3L       668738          A            A  \n",
       "4    oregonr          chr3L       694885          G            G  \n",
       "5    oregonr          chr3L       699318          C            C  \n",
       "6    oregonr          chr3L       714112          A            A  \n",
       "7    oregonr          chr3L       714319          G            G  \n",
       "8    oregonr          chr3L       749771          T            T  \n",
       "9    oregonr          chr3L       749936          T            T  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "SELECT * FROM samples_rearranged\n",
    "WHERE var_adjusted != '.'\n",
    "LIMIT 10;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a17a07bf2f6a46b8bb301ec40c2ae53f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "CREATE OR REPLACE TABLE final_samples AS\n",
    "SELECT \n",
    "    string_split(sample, '-')[1] AS condition,\n",
    "    string_split(sample, '-')[2] AS sample_type,\n",
    "    string_split(sample, '-')[3] AS sample_num,\n",
    "    reference,\n",
    "    variant,\n",
    "    chromosome,\n",
    "    int_pos AS position,\n",
    "    string_split(allele_depth, ',')[1] AS ref_reads,\n",
    "    string_split(allele_depth, ',')[2] AS variant_reads,\n",
    "    quality AS QUAL,\n",
    "    genotype AS GT,\n",
    "    depth AS DP\n",
    "    FROM samples_rearranged\n",
    "ORDER BY sample_num, chromosome, position;\n",
    "\n",
    "CHECKPOINT;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>int_pos</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>quality</th>\n",
       "      <th>genotype</th>\n",
       "      <th>depth</th>\n",
       "      <th>allele_depth</th>\n",
       "      <th>new_variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7902</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>234.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7903</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>188.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>64</td>\n",
       "      <td>62</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7904</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>222.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7905</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>222.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7906</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>222.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7907</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>222.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7908</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>222.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>64</td>\n",
       "      <td>64</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7909</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>228.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7910</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>200.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>68</td>\n",
       "      <td>66</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WT-G0-001</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>7911</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>234.996002</td>\n",
       "      <td>0/0</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      sample chromosome  int_pos reference variant     quality genotype  \\\n",
       "0  WT-G0-001      chr2L     7902         A       .  234.996002      0/0   \n",
       "1  WT-G0-001      chr2L     7903         A       .  188.996002      0/0   \n",
       "2  WT-G0-001      chr2L     7904         C       .  222.996002      0/0   \n",
       "3  WT-G0-001      chr2L     7905         T       .  222.996002      0/0   \n",
       "4  WT-G0-001      chr2L     7906         C       .  222.996002      0/0   \n",
       "5  WT-G0-001      chr2L     7907         C       .  222.996002      0/0   \n",
       "6  WT-G0-001      chr2L     7908         C       .  222.996002      0/0   \n",
       "7  WT-G0-001      chr2L     7909         G       .  228.996002      0/0   \n",
       "8  WT-G0-001      chr2L     7910         C       .  200.996002      0/0   \n",
       "9  WT-G0-001      chr2L     7911         G       .  234.996002      0/0   \n",
       "\n",
       "   depth allele_depth new_variant  \n",
       "0     68           68           A  \n",
       "1     64           62           A  \n",
       "2     64           64           C  \n",
       "3     64           64           T  \n",
       "4     64           64           C  \n",
       "5     64           64           C  \n",
       "6     64           64           C  \n",
       "7     66           66           G  \n",
       "8     68           66           C  \n",
       "9     68           68           G  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT * FROM vcfs\n",
    "WHERE chromosome = 'chr2L' AND int_pos >= 7902\n",
    "LIMIT 10;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>sample_num</th>\n",
       "      <th>reference</th>\n",
       "      <th>variant</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>ref_reads</th>\n",
       "      <th>variant_reads</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>GT</th>\n",
       "      <th>DP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6530334</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6530561</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6531387</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6531408</td>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6532212</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6532213</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6532219</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6532277</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6534904</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>WT</td>\n",
       "      <td>G0</td>\n",
       "      <td>003</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chrX</td>\n",
       "      <td>6534909</td>\n",
       "      <td>0</td>\n",
       "      <td>97</td>\n",
       "      <td>221.998993</td>\n",
       "      <td>1/1</td>\n",
       "      <td>97</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  condition sample_type sample_num reference variant chromosome  position  \\\n",
       "0        WT          G0        003         T       C       chrX   6530334   \n",
       "1        WT          G0        003         A       G       chrX   6530561   \n",
       "2        WT          G0        003         G       A       chrX   6531387   \n",
       "3        WT          G0        003         C       G       chrX   6531408   \n",
       "4        WT          G0        003         A       G       chrX   6532212   \n",
       "5        WT          G0        003         G       A       chrX   6532213   \n",
       "6        WT          G0        003         G       T       chrX   6532219   \n",
       "7        WT          G0        003         G       T       chrX   6532277   \n",
       "8        WT          G0        003         G       A       chrX   6534904   \n",
       "9        WT          G0        003         C       T       chrX   6534909   \n",
       "\n",
       "  ref_reads variant_reads        QUAL   GT   DP  \n",
       "0         0            72  221.998993  1/1   72  \n",
       "1         0            79  221.998993  1/1   79  \n",
       "2         0            88  221.998993  1/1   88  \n",
       "3         0            81  221.998993  1/1   81  \n",
       "4         0            79  221.998993  1/1   79  \n",
       "5         0            79  221.998993  1/1   79  \n",
       "6         0            68  221.998993  1/1   68  \n",
       "7         0            48  221.998993  1/1   48  \n",
       "8         0           105  221.998993  1/1  105  \n",
       "9         0            97  221.998993  1/1   97  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "SELECT * FROM final_samples\n",
    "LIMIT 10;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "vscode": {
     "languageId": "sql"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Success</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Success]\n",
       "Index: []"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "COPY (SELECT * FROM final_samples)\n",
    "TO '{{ref_out_file}}'\n",
    "(FORMAT 'parquet');\n",
    "\n",
    "COPY (SELECT chromosome, int_pos AS position, ref_allele AS reference, alt_allele AS variant FROM ref)\n",
    "TO '{{sample_out_file}}'\n",
    "(FORMAT 'parquet');\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "duckdb",
   "language": "python",
   "name": "duckdb"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
